todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenter.java

```diff
@@ -141,46 +158,89 @@ public void setFiltering(TasksFilterType requestType) {
 
     @Override
     public void editTask() {
-        mTaskDetailPresenter.editTask();
-        mTasksPresenter.loadTasks(false);
+        assert mTaskDetailPresenter != null; // Only called from detail view
+        mTasksMvpController.editTask(getDetailTaskId());
     }
 
     @Override
     public void deleteTask() {
-        if (mTaskDetailPresenter.getTaskId() != null) {
-            mTasksRepository.deleteTask(mTaskDetailPresenter.getTaskId());
+        assert mTaskDetailPresenter != null; // Only called from detail view
+        if (getDetailTaskId() != null) {
+            mTasksRepository.deleteTask(getDetailTaskId());
         }
-        mTaskDetailPresenter.setTaskId(null); // Show an empty detail view
+        mTaskDetailPresenter.setDetailTaskId(null); // Show an empty detail view
+        mTaskDetailPresenter.loadTask();
         mTasksPresenter.loadTasks(false); // Reload the list
     }
 
     @Override
     public void completeTask() {
+        assert mTaskDetailPresenter != null; // Only called from detail view
         mTaskDetailPresenter.completeTask();
         mTasksPresenter.loadTasks(false);
     }
 
     @Override
     public void activateTask() {
+        assert mTaskDetailPresenter != null; // Only called from detail view
         mTaskDetailPresenter.activateTask();
         mTasksPresenter.loadTasks(false);
     }
 
     @Override
     public void loadTask() {
+        assert mTaskDetailPresenter != null; // Only called from detail view
         mTaskDetailPresenter.loadTask();
     }
 
     @Override
-    public void setTaskId(String taskId) {
-        mTaskDetailPresenter.setTaskId(taskId);
+    public void setDetailTaskId(String taskId) {
+        if(mTaskDetailPresenter != null) {
+            mTaskDetailPresenter.setDetailTaskId(taskId);
+        }
+        mTasksPresenter.setSelectedTaskId(taskId);
+    }
+
+    @Override
+    public String getDetailTaskId() {
+        if (mTaskDetailPresenter != null) {
+            return mTaskDetailPresenter.getDetailTaskId();
+        }
+        return null;
     }
 
+    /* AddEditTaskContract.Presenter methods */
+
     @Override
-    public String getTaskId() {
-        if (mTaskDetailPresenter == null) {
-            return null;
+    public void saveTask(String title, String description) {
+        if (mAddEditPresenter != null) {
+            mAddEditPresenter.saveTask(title, description);
+        }
+        mTasksPresenter.onTaskAdded();
```

Comment:
Calling `TasksPresenter.onTaskAdded();` in this way could result in showing a wrong 'New TODO saved' toast.  What if the title and the description of the new TODO are empty? The confirmation toast will be shown anyway, even when the task has not been saved.

This can be fixed calling `onTaskAdded()` of `TaskPresenter` inside [`createTask()`](https://github.com/JoseAlcerreca/android-architecture/blob/61760a9410b5003f6b167e084ff1bbc4094d4719/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskPresenter.java#L114-L114)


https://vid.me/T409